src/aiod/authentication/authentication.py

```diff
@@ -0,0 +1,113 @@
+import http.client
+import requests
+from keycloak import KeycloakOpenID, KeycloakAuthenticationError, KeycloakPostError
+from typing import Sequence, NamedTuple
+
+from aiod.configuration import config
+
+
+def _connect_keycloak() -> KeycloakOpenID:
+    return KeycloakOpenID(
+        server_url=config.auth_server_url,
+        client_id=config.client_id,
+        realm_name=config.realm,
+    )
+
+
+def keycloak_openid() -> KeycloakOpenID:
+    global _keycloak_openid
+    if _keycloak_openid is None:
+        _keycloak_openid = _connect_keycloak()
+    return _keycloak_openid
+
+
+def on_keycloak_config_changed(_: str, __: str, ___: str) -> None:
+    global _keycloak_openid
+    logout(ignore_post_error=True)
+    _keycloak_openid = None
+
+
+config.subscribe("auth_server_url", on_change=on_keycloak_config_changed)
+config.subscribe("realm", on_change=on_keycloak_config_changed)
+config.subscribe("client_id", on_change=on_keycloak_config_changed)
+
+_keycloak_openid: KeycloakOpenID | None = None
+
+
+class User(NamedTuple):
+    name: str
+    roles: Sequence[str]
+
+
+def login(username: str, password: str) -> None:
+    """Logs in the user with the provided username and password.
+
+    Args:
+        username (str): The username of the user.
+        password (str): The password of the user.
+
+    Raises:
+        FailedAuthenticationError: If the username or password is missing or wrong.
+    """
+    if username is None or password is None:
+        raise FailedAuthenticationError(
+            "Username and/or password missing! Please provide your credentials and try again."
+        )
```

Comment:
```suggestion
    Raises:
        ValueError: When username or password is empty.
        FailedAuthenticationError: If the username or password is missing or wrong.
    """
    if not username:
        raise ValueError(f"{username!r} is not a valid username, must be a non-empty string.')
    if not password:
        raise ValueError(f"{password!r} is not a valid password, must be a non-empty string.')
```

Assuming they cannot be empty. If empty strings are allowed and you literally want to check for `None` only, it should be a `TypeError`. These are types of errors which are well defined and have standard expected error types.

Reply:
Why did you use `!r` string format here?

Reply:
It automatically adds quotes to the string. This is better than embedding the quotes on the f-string, since in that case also when e.g. `None` is passed there will be quotes.